### PR TITLE
1.15/1938 message when scanning vouchers vvv eee queue

### PR DESCRIPTION
**Pull request by charlesstrange2**
OK! 

so this is a bit mad, but I thought I'd better get this in before breaking it further

there's config about queue setup

I've abstracted and sensible-ised the mess that was the VoucherController::transition() function into 

- a dedicated class for doing that
- with functions that know how do do each of the types of transition

this enables the following request->queue->result stuff.

(POST `/vouchers` with transition data _still works_ for stodgy slow synchronous stuff.)

POST `/vouchers/transition` will start a queued job to do the same work, but return a 202 to the client to tell them where to monitor the job

GET `/queue/{id}` is that endpoint that monitors a queue job. when it finishes it 303 redirects the client to a cached copy of whatever they asked for.

### CODE REVIEW

File: app/Http/Controllers/API/VoucherController.php [hunk 1]
```diff
@@ -2,176 +2,35 @@
...
+use App\Services\TransitionProcessor;
 use App\Trader;
 use App\Voucher;
-use Auth;
-use Carbon\Carbon;
 use Illuminate\Http\JsonResponse;
-use Symfony\Component\Lock\LockFactory;
-use Symfony\Component\Lock\Store\SemaphoreStore;
 
 class VoucherController extends Controller
```
Comment (charlesstrange2) [inline app/Http/Controllers/API/VoucherController.php:12]: this file is basically gutted; most logic went into the TransitionProcess.

File: app/Jobs/ProcessTransitionJob.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Support\Facades\Cache;
+use Illuminate\Support\Str;
+use Imtigger\LaravelJobStatus\JobStatus;
+use Imtigger\LaravelJobStatus\Trackable;
+use Log;
+
+class ProcessTransitionJob implements ShouldQueue
+{
+    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels, Trackable;
```
Comment (charlesstrange2) [inline app/Jobs/ProcessTransitionJob.php:21]: note trackable allows the jobstatus plugin to find this job, something L9 doesn't normally support

File: config/job-status.php [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+return [
+    'model' => \Imtigger\LaravelJobStatus\JobStatus::class,
```
Comment (charlesstrange2) [inline config/job-status.php:4]: defaults from pakage

File: config/queue.php [hunk 4]
```diff
@@ -39,8 +39,11 @@
             'table' => 'jobs',
             'queue' => 'default',
             'retry_after' => 90,
+            // waits for transactions before running jobs
+            'after_commit' => true,
```
Comment (charlesstrange2) [inline config/queue.php:43]: helps if we happy to use a transactioned query.

File: database/migrations/2023_03_10_115250_create_jobs_table.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
```
Comment (charlesstrange2) [inline database/migrations/2023_03_10_115250_create_jobs_table.php:14]: default from l9

**Review by charlesstrange2 — COMMENTED** [review]
not sure why it made so many gitignores.

File: app/Jobs/ProcessTransitionJob.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    {
+        // this is the body data; needs to tell the client what to do
+        $data = array_merge(
+            [
+                'location' => route('api.queued-task.show', ['jobStatus' => $jobStatus->id]),
+                'retry-after' => 2,
+            ],
+            $jobStatus->only(['id', 'status'])
+        );
+        Log::info("hello");
```
Comment (neonleann) [inline app/Jobs/ProcessTransitionJob.php:55]: You forgot our little hello!

File: app/Services/TransitionProcessor.php [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                case 'reject':
+                    $this->handleReject();
+                    break;
+                default:
+                    $this->handleDefault();
+            }
+
+            $lock->release();
+        } else {
+            Log::info('No lock for me!');
```
Comment (neonleann) [inline app/Services/TransitionProcessor.php:80]: Is this deliberate, so we can see the error in production?

File: app/Services/TransitionProcessor.php [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    /**
+     * @return void
+     */
+    public function handleCollect(): void
+    {
+        // Fetch the date we start to care about deliveries
+        $collect_delivery_date = Carbon::parse(config('arc.first_delivery_date'));
+        $transition = $this->transition;
+
+        sleep(10);
```
Comment (neonleann) [inline app/Services/TransitionProcessor.php:95]: Is this still needed now there's a lock?

File: config/queue.php [hunk 9]
```diff
@@ -39,8 +39,11 @@
             'table' => 'jobs',
             'queue' => 'default',
             'retry_after' => 90,
```
Comment (neonleann) [inline config/queue.php:41]: Is this definietly in seconds? I'm just thinking about our over retry disaster.

File: app/Jobs/ProcessTransitionJob.php (hunk 6 above)
Comment (charlesstrange2) [reply app/Jobs/ProcessTransitionJob.php:55]: doh!

File: app/Services/TransitionProcessor.php (hunk 8 above)
Comment (charlesstrange2) [reply app/Services/TransitionProcessor.php:95]: doh!

File: config/queue.php (hunk 9 above)
Comment (charlesstrange2) [reply config/queue.php:41]: yep, it's a default in seconds.

File: app/Services/TransitionProcessor.php (hunk 7 above)
Comment (charlesstrange2) [reply app/Services/TransitionProcessor.php:80]: yeah, could probably be  a better message.

**Review by neonleann — APPROVED** [review]
This looks like it was a lot of hard work. Well done! 👏

**Review by lauragurney — APPROVED** [review]
Oof, what a beast. All looks relatively sensible though 👍

**Pull request merged**